File: api/src/test/java/rw/test/AbstractControllerTest.java
package rw.test;

import org.junit.Before;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.test.IntegrationTest;
import org.springframework.boot.test.SpringApplicationConfiguration;
import org.springframework.boot.test.TestRestTemplate;
import org.springframework.test.context.web.WebAppConfiguration;
import org.springframework.web.client.RestTemplate;
import rw.Application;

import java.net.URL;

@SpringApplicationConfiguration(classes = Application.class)
@WebAppConfiguration
@IntegrationTest({"server.port=0"})
public abstract class AbstractControllerTest {

    @Value("${local.server.port}")
    private int port;

    private URL base;
    private RestTemplate template;

    @Before
    public void setUp() throws Exception {
        this.base = new URL("http://localhost:" + port + "/");
        template = new TestRestTemplate();
    }

    public RestTemplate getTemplate() {
        return template;
    }

    public String getBase() {
        return base.toString();
    }
}

File: api/src/test/java/rw/test/AbstractControllerTest.scala
package rw.test


import java.net.{InetSocketAddress, Proxy}

import com.typesafe.scalalogging.LazyLogging
import org.springframework.beans.factory.annotation.{Autowired, Value}
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.http.client.SimpleClientHttpRequestFactory

@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
abstract class AbstractControllerTest extends LazyLogging{

  @Autowired var template: TestRestTemplate = _
  val proxyHost: String = "localhost"
  val proxyPort: Int = 8888

  @Value("${rest.proxy}")
  var proxy: Boolean = _

  def getTemplate: TestRestTemplate = {
    if (proxy) {
      logger.warn(s"Using proxy on host $proxyHost:$proxyHost")
      template.getRestTemplate.setRequestFactory(getFactory)
    }
    template
  }

  def getFactory = {
    val requestFactory: SimpleClientHttpRequestFactory = new SimpleClientHttpRequestFactory()
    // Proxy via web fiddler, this will allow you to record/replay rest request
    val fiddler: Proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(proxyHost, proxyPort))
    requestFactory.setProxy(fiddler)
    requestFactory
  }


}